### PR TITLE
Add support for tools in Anthropic

### PR DESCRIPTION
**Pull request by flvndvd**
## Description

This PR introduces support for tools in our Anthropic implementation, following the guidelines in this [documentation](https://docs.anthropic.com/claude/docs/tool-use-examples). Note that tools are not yet supported in stream mode due to the beta's limitations. Consequently, we've end up with many types to accommodate all our use cases
.
Anthropic's responses have a flat structure, with the primary distinction between a regular answer and a tool usage being the `type` field. Thanks to serde, we can efficiently manage this difference 👌 .

This PR focuses only on supporting tools in the message API, but it does not yet enable passing tools from previous messages in the messages field. We'll wait for more information about the API before updating this behavior in a separate PR.

### CODE REVIEW

File: core/src/providers/anthropic.rs [hunk 1]
```diff
@@ -75,6 +76,86 @@ pub struct AnthropicContent {
...
+        let arguments = serde_json::to_string(&tool_use.input)?;
+        Ok(ChatFunctionCall {
+            name: tool_use.name.clone(),
+            arguments,
+        })
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[serde(rename_all = "snake_case", tag = "type")]
```
Comment (fontanierh) [inline core/src/providers/anthropic.rs:99]: the tag thing is 🤯

File: core/src/providers/anthropic.rs [hunk 2]
```diff
@@ -114,22 +195,118 @@ impl TryFrom<&ChatMessage> for AnthropicChatMessage {
...
+
+    fn try_from(cr: StreamChatResponse) -> Result<Self, Self::Error> {
+        let content = cr
+            .content
+            .into_iter()
+            .map(AnthropicResponseContent::try_from)
+            .collect::<Result<Vec<AnthropicResponseContent>, anyhow::Error>>()?;
+
+        Ok(ChatResponse {
+            id: cr.id.clone(),
```
Comment (fontanierh) [inline core/src/providers/anthropic.rs:262]: Do you really need the clones here ?

Comment (fontanierh) [reply core/src/providers/anthropic.rs:262]: feels like `AnthropicResponseContent` is owned no ?

File: core/src/providers/anthropic.rs [hunk 3]
```diff
@@ -304,6 +485,11 @@ impl AnthropicLLM {
     ) -> Result<ChatResponse> {
         assert!(self.api_key.is_some());
 
+        // Streaming (stream=true) is not yet supported on tools.
+        if tools.len() > 0 {
+            return Err(anyhow!("Anthropic does not support chat functions."));
```
Comment (spolu) [inline core/src/providers/anthropic.rs:490]: Message is off?

**Review by spolu — APPROVED** [review]
Looks very reasonable!

File: core/src/providers/anthropic.rs [hunk 4]
```diff
@@ -114,22 +195,118 @@ impl TryFrom<&ChatMessage> for AnthropicChatMessage {
...
+// This code converts a ChatResponse to a ChatMessage, but only supports one tool call.
+// It takes the first tool call from the vector of AnthropicResponseContent,
+// potentially discarding others. Anthropic often returns the CoT content as a first message,
+// which gets combined with the first tool call in the resulting ChatMessage.
+impl TryFrom<ChatResponse> for ChatMessage {
+    type Error = anyhow::Error;
+
+    fn try_from(cr: ChatResponse) -> Result<Self, Self::Error> {
+        let text_content = cr.content.iter().find_map(|item| match item.get_text() {
+            Some(text) => Some(text.clone()),
```
Comment (fontanierh) [inline core/src/providers/anthropic.rs:281]: do you need to clone here ? Can you clone closer to when you actually need it so it's more obvious why ?

File: core/src/providers/anthropic.rs [hunk 5]
```diff
@@ -114,22 +195,118 @@ impl TryFrom<&ChatMessage> for AnthropicChatMessage {
...
+            _ => None,
+        });
+
+        let tool_uses: Vec<ToolUse> = cr
+            .content
+            .iter()
+            .filter_map(|item| match item.get_tool_use() {
+                Some(tool_use) => Some(tool_use.clone()),
```
Comment (fontanierh) [inline core/src/providers/anthropic.rs:289]: I don't think this clone is needed

File: core/src/providers/anthropic.rs [hunk 6]
```diff
@@ -256,11 +434,16 @@ impl AnthropicLLM {
...
+        if tools.len() > 0 {
+            body["tools"] = json!(tools);
+        }
+
         let res = reqwest::Client::new()
             .post(self.messages_uri()?.to_string())
             .header("Content-Type", "application/json")
             .header("X-API-Key", self.api_key.clone().unwrap())
             .header("anthropic-version", "2023-06-01")
+            .header("anthropic-beta", "tools-2024-04-04")
```
Comment (fontanierh) [inline core/src/providers/anthropic.rs:446]: let's only add if !tools.is_empty

File: core/src/providers/anthropic.rs [hunk 7]
```diff
@@ -256,11 +434,16 @@ impl AnthropicLLM {
             body["system"] = json!(system);
         }
 
+        if tools.len() > 0 {
```
Comment (fontanierh) [inline core/src/providers/anthropic.rs:437]: nit: is_empty()

File: core/src/providers/anthropic.rs [hunk 8]
```diff
@@ -304,6 +485,11 @@ impl AnthropicLLM {
     ) -> Result<ChatResponse> {
         assert!(self.api_key.is_some());
 
+        // Streaming (stream=true) is not yet supported on tools.
+        if tools.len() > 0 {
```
Comment (fontanierh) [inline core/src/providers/anthropic.rs:489]: nit is_empty

**Review by fontanierh — COMMENTED** [review]
Looks great, can we try to remove a few clones ?

**Review by fontanierh — APPROVED** [review]
LGTM

**Pull request merged**